### PR TITLE
add a function for imgcat + url

### PR DESCRIPTION
**Pull request by ShuaiHsunLee**
curl
I utilize curl to check if the file can be downloaded from the provided url.
If yes, downloads it in the path "~/.iterm2/image_collection" and create a list.txt.
list.txt file is used to check if the user has downloads file from the url.
Users can save time if they have already downloaded the file from the url.

md5
I use the value of each file's md5 as its name.
If I download a new file whose md5 value is identical to the file user downloaded, in practical, the new file is the same file.
Therefore, we only need to store one.

### CODE REVIEW

**Comment by gnachman** [comment]
This is a nice addition, but I think this tool should not perform caching. It's unexpected and has the side effect of using unbounded amounts of disk space. It's also dangerous to perform network requests, so there should be a clear user intent before invoking curl. I'd suggest doing that only if the command is like `imgcat --url http://example.com/image.gif`.

File: source/utilities/imgcat [hunk 1]
```diff
@@ -86,6 +86,29 @@ while [ $# -gt 0 ]; do
     -p|--p|--print)
         print_filename=1
         ;;
+	-u|--u|--url)
+        if curl --output /dev/null --silent --head --fail "$2"; then # check if the url is downloadable
+			# create a file with unique file name
+			file=$(mktemp -t curlimg) || { echo "Failed creating file"; return; }
+			# download file and save it into the unique file name just created
+    		curl -s "$2" > $file || { echo "Failed to curl file"; return; }
```
Comment (gnachman) [inline source/utilities/imgcat]: `echo "curl failed for url $2"; exit 1`

File: source/utilities/imgcat [hunk 2]
```diff
@@ -86,6 +86,29 @@ while [ $# -gt 0 ]; do
...
+			# reset argument of url to the corresponding unique file
+            has_stdin=f
+            print_image "$file" 1 "$(base64 < "$file")" "$print_filename"
+			# delete the downloaded file
+			rm $file
```
Comment (gnachman) [inline source/utilities/imgcat]: Put $file in "quotes"

File: source/utilities/imgcat [hunk 3]
```diff
@@ -95,7 +118,7 @@ while [ $# -gt 0 ]; do
         if [ -r "$1" ] ; then
             has_stdin=f
             print_image "$1" 1 "$(base64 < "$1")" "$print_filename"
-        else
+		else
```
Comment (gnachman) [inline source/utilities/imgcat]: Please don't use tabs for indentation. Make sure to convert all tabs to spaces.

File: source/utilities/imgcat [hunk 4]
```diff
@@ -86,6 +86,29 @@ while [ $# -gt 0 ]; do
     -p|--p|--print)
         print_filename=1
         ;;
+	-u|--u|--url)
+        if curl --output /dev/null --silent --head --fail "$2"; then # check if the url is downloadable
+			# create a file with unique file name
+			file=$(mktemp -t curlimg) || { echo "Failed creating file"; return; }
```
Comment (gnachman) [inline source/utilities/imgcat]: `echo "mktemp failed for url $2"; exit 1`

File: source/utilities/imgcat [hunk 5]
```diff
@@ -86,6 +86,29 @@ while [ $# -gt 0 ]; do
...
+			# download file and save it into the unique file name just created
+    		curl -s "$2" > $file || { echo "Failed to curl file"; return; }
+			# reset argument of url to the corresponding unique file
+            has_stdin=f
+            print_image "$file" 1 "$(base64 < "$file")" "$print_filename"
+			# delete the downloaded file
+			rm $file
+			# only modify second argument in order to keep running option "-u"
+			# shift will remove the first arguement "-u"
+			set -- ${@:1:1} "-u" ${@:3}
```
Comment (gnachman) [inline source/utilities/imgcat]: Please explain exactly what this command does. I've never seen this syntax before and it's basically impossible to google for.

File: source/utilities/imgcat [hunk 6]
```diff
@@ -86,6 +86,29 @@ while [ $# -gt 0 ]; do
     -p|--p|--print)
         print_filename=1
         ;;
+	-u|--u|--url)
+        if curl --output /dev/null --silent --head --fail "$2"; then # check if the url is downloadable
```
Comment (gnachman) [inline source/utilities/imgcat]: It's kind of ugly to make two requests. I haven't tested this, but I think something along these lines could be made to work:

```
encoded_image=$(curl -s "$2" | base64) || (error "No such file or url $2"; exit 2)
has_stdin=f
print_image "$file" 1 "$encoded_image" "$print_filename"
etc.
```

Comment (ShuaiHsunLee) [reply source/utilities/imgcat]: I found the ugly is from indentations. I will check it out.
Moreover, your code is very precise and elegant. I will change it as well.

Comment (ShuaiHsunLee) [reply source/utilities/imgcat]: set -- ${@:1:1} "-u" ${@:3}
The code is to shift all the argument without $1 and $2.
"-u" replaces $1 in order to make sure while loop will keep running -u option.

if [ "$#" -eq 2 ]; then
        exit
fi
The code is to make sure no url needed to printf picture.

To sum up, this four lines of code is to printf multiple url pictures.

**Comment by gnachman** [comment]
Merged with some small changes. Thanks1

**Pull request closed without merging**